pallets/msa/src/offchain_storage.rs

```diff
@@ -0,0 +1,464 @@
+pub use common_primitives::msa::MessageSourceId;
+/// Offchain Storage for MSA
+use common_primitives::offchain::{
+	self as offchain_common, get_msa_account_lock_name, get_msa_account_storage_key_name,
+	LockStatus, MSA_ACCOUNT_LOCK_TIMEOUT_EXPIRATION,
+};
+use frame_support::RuntimeDebugNoBound;
+use frame_system::pallet_prelude::BlockNumberFor;
+use parity_scale_codec::{Decode, Encode};
+use sp_core::serde::{Deserialize, Serialize};
+extern crate alloc;
+use alloc::string::String;
+use sp_io::offchain_index;
+use sp_runtime::{
+	offchain::{
+		storage::StorageValueRef,
+		storage_lock::{BlockAndTime, StorageLock, Time},
+		Duration,
+	},
+	traits::One,
+	Saturating,
+};
+use sp_std::{collections::btree_map::BTreeMap, fmt::Debug, vec, vec::Vec};
+
+/// Block event storage prefix
+const BLOCK_EVENT_KEY: &[u8] = b"frequency::block_event::msa::";
+
+/// Offchain index for MSA events count
+const BLOCK_EVENT_COUNT_KEY: &[u8] = b"frequency::block_event::msa::count::";
+
+/// Lock expiration timeout in in milli-seconds for initial data import msa pallet
+const MSA_INITIAL_LOCK_TIMEOUT_EXPIRATION: u64 = 3000;
+
+/// Lock expiration block for initial data import msa pallet
+const MSA_INITIAL_LOCK_BLOCK_EXPIRATION: u32 = 30;
+
+/// Lock name for initial data index for msa pallet
+const MSA_INITIAL_LOCK_NAME: &[u8; 28] = b"Msa::ofw::initial-index-lock";
+
+/// storage name for initial data import storage
+pub const MSA_INITIAL_INDEXED_STORAGE_NAME: &[u8; 25] = b"Msa::ofw::initial-indexed";
+
+/// Lock name for last processed block number events
+const LAST_PROCESSED_BLOCK_LOCK_NAME: &[u8; 35] = b"Msa::ofw::last-processed-block-lock";
+
+/// lst processed block storage name
+pub const LAST_PROCESSED_BLOCK_STORAGE_NAME: &[u8; 30] = b"Msa::ofw::last-processed-block";
+
+/// Lock expiration timeout in in milli-seconds for last processed block
+const LAST_PROCESSED_BLOCK_LOCK_TIMEOUT_EXPIRATION: u64 = 5000;
+
+/// Lock expiration for last processed block
+const LAST_PROCESSED_BLOCK_LOCK_BLOCK_EXPIRATION: u32 = 20;
+
+/// number of previous blocks to check to mitigate offchain worker skips processing any block
+const NUMBER_OF_PREVIOUS_BLOCKS_TO_CHECK: u32 = 5u32;
+
+/// number of blocks to explore when trying to find the block number from block hash
+const NUMBER_OF_BLOCKS_TO_EXPLORE: u32 = 1000;
+
+/// HTTP request deadline in milliseconds
+pub const HTTP_REQUEST_DEADLINE_MILLISECONDS: u64 = 2000;
+
+/// LOCAL RPC URL and port
+/// warning: this should be updated if rpc port is set to anything different from 9944
+pub const RPC_FINALIZED_BLOCK_REQUEST_URL: &'static str = "http://localhost:9944";
+/// request body for getting last finalized block from rpc
+pub const RPC_FINALIZED_BLOCK_REQUEST_BODY: &[u8; 78] =
+	b"{\"id\": 10, \"jsonrpc\": \"2.0\", \"method\": \"chain_getFinalizedHead\", \"params\": []}";
+
+/// stores the event into offchain DB using offchain indexing
+pub fn offchain_index_event<T: Config>(event: &Event<T>, msa_id: MessageSourceId) {
```

Comment:
I'd be for one of those. Worth a discussion on what is the better model:
* One generic `MsaKeyModified` event:
    * Pro: It's generic :smile:
    * Con: bigger footprint in the block (added enum) (but probably only the size of a `u8`)
    * Con(?): Impacts more extrinsics
    * Con(?): clients have to decode the entire event payload to determine the action, if they're filtering on Add/Update/Delete.
* Just do `PublicKeyDeletedV2`
    * Pro: Limited scope of changes to chain code
    * Pro: Minimal client impact
    * Con(?): Generic event might be better design-wise
    
(of course, neither is strictly necessary, works as-is)

Reply:
I think having the generic event idea effort is outside the scope of this PR.
One reason to keep the type of regular events separate from indexed events is that in case we would like to do offchain indexing on other things we can not always guarantee all the data would be in the event and I don't think it's an extendable approach.